Share SIL TypeConverters between SILModules in batch mode

slavapestov opened the pull request:
Batch mode compiles a single AST ModuleDecl but creates a new SILModule for each translation unit emitted by SILGen/IRGen. This SILModule had a lot of cached state that did not depend on the translation unit and was being re-created for each source file.

This PR decouples the TypeConverter from the SILModule, making it only depend on the ModuleDecl instead, and refactors the frontend to share a single instance across all SILModules it creates.

slavapestov committed:
SIL: Refactor TypeConverter to not require a SILModule

This mostly requires changing various entry points to pass around a
TypeConverter instead of a SILModule. I've left behind entry points
that take a SILModule for a few methods like SILType::subst() to
avoid creating even more churn.

slavapestov committed:
SIL: Share TypeConverter between SILModules in batch mode

eeckstein reviewed (approved):
nice! Did you see any compile time/memory impact?

eeckstein commented on include/swift/SIL/SILType.h:
```diff
@@ -228,26 +228,29 @@ class SILType {
-                            CanGenericSignature Sig,
-                            ResilienceExpansion Expansion);
+  /// tc.getTypeLowering(type).isAddressOnly().
+  static bool isAddressOnly(CanType type, Lowering::TypeConverter &tc,
```
changing the API from SILModule to TypeConverter is not strictly required, right? isAddressOnly could still get the type converter reference from the module.

eeckstein replied:
ok, makes sense

Bx-Flaco86 commented:
Ok

On Wed, Aug 28, 2019 at 12:15 PM swift-ci <notifications@github.com> wrote:

> Compilation-performance test failed
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apple/swift/pull/26897?email_source=notifications&email_token=ANAIUEGIL4CAIOQYX5B7T4TQG2QA7A5CNFSM4IQVW6G2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5LU6AI#issuecomment-525815553>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ANAIUEFFO7PQHF6HYJ7BBSTQG2QA7ANCNFSM4IQVW6GQ>
> .
>

slavapestov committed:
Frontend: Only build and generate a single SILModule at a time

Instead of SILGen'ing all primary files before we go on to optimize
and IRGen them, run each file to completion before starting the next
one. This reduces memory usage.

slavapestov commented:
# Summary for master full

**Unexpected test results, excluded stats for RxCocoa, Base64CoderSwiftUI, SwifterSwift**

**Regressions found (see below)**

- [Debug-batch](#debug-batch-49cfcd52a955fa2d)
    - [debug-batch brief](#debug-batch-brief-49cfcd52a955fa2d)
    - [debug-batch detailed](#debug-batch-detailed-49cfcd52a955fa2d)
- [Release](#release-49cfcd52a955fa2d)
    - [release brief](#release-brief-49cfcd52a955fa2d)
    - [release detailed](#release-detailed-49cfcd52a955fa2d)



<a name="debug-batch-49cfcd52a955fa2d">

# Debug-batch

<a name="debug-batch-brief-49cfcd52a955fa2d">

## debug-batch brief
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (3)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 46,790,901,943,365 | 47,036,708,379,140 | 245,806,435,775 | 0.53%
LLVM.NumLLVMBytesOutput | 1,793,882,048 | 1,793,884,868 | 2,820 | 0.0%
time.swift-driver.wall | 5159.1s | 5157.4s | -1.7s | -0.03%
</details>

<a name="debug-batch-detailed-49cfcd52a955fa2d">

## debug-batch detailed
<details>
<summary>Regressed (5)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumASTBytesAllocated | 62,551,489,068 | 63,347,406,727 | 795,917,659 | 1.27% :no_entry:
Sema.AccessLevelRequest | 11,396,059 | 11,639,521 | 243,462 | 2.14% :no_entry:
Sema.CollectOverriddenDeclsRequest | 6,936,132 | 7,080,938 | 144,806 | 2.09% :no_entry:
Sema.ProvideDefaultImplForRequest | 6,936,132 | 7,080,938 | 144,806 | 2.09% :no_entry:
Sema.USRGenerationRequest | 8,155,226 | 8,296,047 | 140,821 | 1.73% :no_entry:
</details>
<details>
<summary>Improved (2)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Driver.NumDriverPipePolls | 59,731 | 58,416 | -1,315 | -2.2% :white_check_mark:
Driver.NumDriverPipeReads | 46,732 | 45,422 | -1,310 | -2.8% :white_check_mark:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (154)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumASTScopeLookups | 0 | 0 | 0 | 0.0%
AST.NumBraceStmtASTScopeExpansions | 0 | 0 | 0 | 0.0%
AST.NumBraceStmtASTScopes | 0 | 0 | 0 | 0.0%
AST.NumDecls | 140,002 | 140,002 | 0 | 0.0%
AST.NumDependencies | 321,521 | 321,519 | -2 | -0.0%
AST.NumInfixOperators | 53,779 | 53,779 | 0 | 0.0%
AST.NumIterableTypeBodyASTScopeExpansions | 0 | 0 | 0 | 0.0%
AST.NumIterableTypeBodyASTScopes | 0 | 0 | 0 | 0.0%
AST.NumLinkLibraries | 0 | 0 | 0 | 0.0%
AST.NumLoadedModules | 399,488 | 399,488 | 0 | 0.0%
AST.NumLocalTypeDecls | 253 | 253 | 0 | 0.0%
AST.NumLookupInModule | 6,757,318 | 6,740,971 | -16,347 | -0.24%
AST.NumLookupQualifiedInAnyObject | 291 | 291 | 0 | 0.0%
AST.NumLookupQualifiedInModule | 3,268,494 | 3,252,021 | -16,473 | -0.5%
AST.NumLookupQualifiedInNominal | 8,123,033 | 8,137,467 | 14,434 | 0.18%
AST.NumModuleLookupClassMember | 16,334 | 16,334 | 0 | 0.0%
AST.NumModuleLookupValue | 45,570,476 | 45,549,632 | -20,844 | -0.05%
AST.NumObjCMethods | 24,701 | 24,701 | 0 | 0.0%
AST.NumPostfixOperators | 23 | 23 | 0 | 0.0%
AST.NumPrecedenceGroups | 25,668 | 25,668 | 0 | 0.0%
AST.NumPrefixOperators | 99 | 99 | 0 | 0.0%
AST.NumReferencedDynamicNames | 199 | 199 | 0 | 0.0%
AST.NumReferencedMemberNames | 6,335,202 | 6,335,202 | 0 | 0.0%
AST.NumReferencedTopLevelNames | 476,992 | 476,992 | 0 | 0.0%
AST.NumSourceBuffers | 566,010 | 566,010 | 0 | 0.0%
AST.NumSourceLines | 4,685,680 | 4,685,680 | 0 | 0.0%
AST.NumSourceLinesPerSecond | 3,295,401 | 3,300,031 | 4,630 | 0.14%
AST.NumTotalClangImportedEntities | 6,439,331 | 6,442,093 | 2,762 | 0.04%
AST.NumUnqualifiedLookup | 4,379,399 | 4,379,576 | 177 | 0.0%
Driver.ChildrenMaxRSS | 236,419,827,712 | 237,031,981,056 | 612,153,344 | 0.26%
Driver.DriverDepCascadingDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepCascadingTopLevel | 0 | 0 | 0 | 0.0%
Driver.DriverDepDynamic | 0 | 0 | 0 | 0.0%
Driver.DriverDepExternal | 0 | 0 | 0 | 0.0%
Driver.DriverDepMember | 0 | 0 | 0 | 0.0%
Driver.DriverDepNominal | 0 | 0 | 0 | 0.0%
Driver.DriverDepTopLevel | 0 | 0 | 0 | 0.0%
Driver.NumDriverJobsRun | 27,904 | 27,904 | 0 | 0.0%
Driver.NumDriverJobsSkipped | 0 | 0 | 0 | 0.0%
Driver.NumProcessFailures | 0 | 0 | 0 | 0.0%
Frontend.MaxMallocUsage | 1,327,826,434,312 | 1,331,125,290,832 | 3,298,856,520 | 0.25%
Frontend.NumInstructionsExecuted | 46,790,901,943,365 | 47,036,708,379,140 | 245,806,435,775 | 0.53%
Frontend.NumProcessFailures | 0 | 0 | 0 | 0.0%
IRModule.NumIRAliases | 197,196 | 197,196 | 0 | 0.0%
IRModule.NumIRBasicBlocks | 6,834,083 | 6,834,083 | 0 | 0.0%
IRModule.NumIRComdatSymbols | 0 | 0 | 0 | 0.0%
IRModule.NumIRFunctions | 3,374,319 | 3,374,319 | 0 | 0.0%
IRModule.NumIRGlobals | 3,535,040 | 3,535,040 | 0 | 0.0%
IRModule.NumIRIFuncs | 0 | 0 | 0 | 0.0%
IRModule.NumIRInsts | 87,091,678 | 87,091,678 | 0 | 0.0%
IRModule.NumIRNamedMetaData | 134,220 | 134,220 | 0 | 0.0%
IRModule.NumIRValueSymbols | 6,254,504 | 6,254,504 | 0 | 0.0%
LLVM.NumLLVMBytesOutput | 1,793,882,048 | 1,793,884,868 | 2,820 | 0.0%
Parse.NumFunctionsParsed | 267,975 | 267,975 | 0 | 0.0%
Parse.NumIterableDeclContextParsed | 908,893 | 908,946 | 53 | 0.01%
SILModule.NumSILGenDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILGenFunctions | 1,709,502 | 1,709,502 | 0 | 0.0%
SILModule.NumSILGenGlobalVariables | 53,884 | 53,884 | 0 | 0.0%
SILModule.NumSILGenVtables | 18,576 | 18,576 | 0 | 0.0%
SILModule.NumSILGenWitnessTables | 72,169 | 72,169 | 0 | 0.0%
SILModule.NumSILOptDefaultWitnessTables | 0 | 0 | 0 | 0.0%
SILModule.NumSILOptFunctions | 2,450,953 | 2,450,953 | 0 | 0.0%
SILModule.NumSILOptGlobalVariables | 55,492 | 55,492 | 0 | 0.0%
SILModule.NumSILOptVtables | 31,332 | 31,332 | 0 | 0.0%
SILModule.NumSILOptWitnessTables | 157,761 | 157,761 | 0 | 0.0%
Sema.AbstractGenericSignatureRequest | 26,211 | 26,211 | 0 | 0.0%
Sema.AttachedFunctionBuilderRequest | 3 | 3 | 0 | 0.0%
Sema.AttachedPropertyWrapperTypeRequest | 515,933 | 515,933 | 0 | 0.0%
Sema.AttachedPropertyWrappersRequest | 2,173,995 | 2,173,996 | 1 | 0.0%
Sema.ClassAncestryFlagsRequest | 109,613 | 109,613 | 0 | 0.0%
Sema.CursorInfoRequest | 0 | 0 | 0 | 0.0%
Sema.CustomAttrNominalRequest | 3 | 3 | 0 | 0.0%
Sema.DefaultAndMaxAccessLevelRequest | 57,729 | 57,740 | 11 | 0.02%
Sema.DefaultDefinitionTypeRequest | 8,104 | 8,104 | 0 | 0.0%
Sema.DefaultTypeRequest | 461,168 | 461,168 | 0 | 0.0%
Sema.EmittedMembersRequest | 27,953 | 27,953 | 0 | 0.0%
Sema.EnumRawTypeRequest | 25,102 | 25,102 | 0 | 0.0%
Sema.ExistentialConformsToSelfRequest | 23,243 | 23,298 | 55 | 0.24%
Sema.ExistentialTypeSupportedRequest | 17,731 | 17,731 | 0 | 0.0%
Sema.ExtendedNominalRequest | 6,850,362 | 6,863,837 | 13,475 | 0.2%
Sema.ExtendedTypeRequest | 115,702 | 115,716 | 14 | 0.01%
Sema.FunctionBuilderTypeRequest | 3 | 3 | 0 | 0.0%
Sema.FunctionOperatorRequest | 76,837 | 76,837 | 0 | 0.0%
Sema.GetDestructorRequest | 28,360 | 28,360 | 0 | 0.0%
Sema.HasDynamicMemberLookupAttributeRequest | 0 | 0 | 0 | 0.0%
Sema.InheritedDeclsReferencedRequest | 6,194,931 | 6,203,351 | 8,420 | 0.14%
Sema.InheritedTypeRequest | 330,517 | 330,464 | -53 | -0.02%
Sema.InitKindRequest | 112,507 | 112,507 | 0 | 0.0%
Sema.IsAccessorTransparentRequest | 320,252 | 320,252 | 0 | 0.0%
Sema.IsDeclApplicableRequest | 0 | 0 | 0 | 0.0%
Sema.IsDynamicRequest | 1,664,633 | 1,664,633 | 0 | 0.0%
Sema.IsFinalRequest | 2,588,235 | 2,607,865 | 19,630 | 0.76%
Sema.IsGetterMutatingRequest | 443,267 | 443,267 | 0 | 0.0%
Sema.IsImplicitlyUnwrappedOptionalRequest | 2,532,862 | 2,533,120 | 258 | 0.01%
Sema.IsObjCRequest | 1,597,289 | 1,600,573 | 3,284 | 0.21%
Sema.IsSetterMutatingRequest | 350,119 | 350,119 | 0 | 0.0%
Sema.LazyStoragePropertyRequest | 2,603 | 2,603 | 0 | 0.0%
Sema.MangleLocalTypeDeclRequest | 506 | 506 | 0 | 0.0%
Sema.NamedLazyMemberLoadFailureCount | 22,402 | 22,359 | -43 | -0.19%
Sema.NamedLazyMemberLoadSuccessCount | 29,873,363 | 29,852,388 | -20,975 | -0.07%
Sema.NominalTypeLookupDirectCount | 36,407,153 | 36,484,070 | 76,917 | 0.21%
Sema.NumAccessorBodiesSynthesized | 189,296 | 189,296 | 0 | 0.0%
Sema.NumAccessorsSynthesized | 296,229 | 296,229 | 0 | 0.0%
Sema.NumConformancesDeserialized | 9,338,086 | 9,416,332 | 78,246 | 0.84%
Sema.NumConstraintScopes | 27,055,883 | 27,065,340 | 9,457 | 0.03%
Sema.NumConstraintsConsideredForEdgeContraction | 86,080,949 | 86,081,181 | 232 | 0.0%
Sema.NumCyclicOneWayComponentsCollapsed | 0 | 0 | 0 | 0.0%
Sema.NumDeclsDeserialized | 74,294,165 | 74,850,945 | 556,780 | 0.75%
Sema.NumDeclsTypechecked | 1,403,528 | 1,403,528 | 0 | 0.0%
Sema.NumDeclsValidated | 2,589,263 | 2,589,321 | 58 | 0.0%
Sema.NumFunctionsTypechecked | 527,256 | 527,256 | 0 | 0.0%
Sema.NumGenericSignatureBuilders | 1,459,994 | 1,466,700 | 6,706 | 0.46%
Sema.NumLazyGenericEnvironments | 13,166,208 | 13,296,038 | 129,830 | 0.99%
Sema.NumLazyGenericEnvironmentsLoaded | 312,987 | 312,843 | -144 | -0.05%
Sema.NumLazyIterableDeclContexts | 9,684,379 | 9,703,582 | 19,203 | 0.2%
Sema.NumLazyRequirementSignatures | 1,024,901 | 1,026,013 | 1,112 | 0.11%
Sema.NumLazyRequirementSignaturesLoaded | 660,575 | 661,327 | 752 | 0.11%
Sema.NumLeafScopes | 17,402,841 | 17,411,313 | 8,472 | 0.05%
Sema.NumTypesDeserialized | 22,262,897 | 22,382,841 | 119,944 | 0.54%
Sema.NumTypesValidated | 1,817,214 | 1,817,252 | 38 | 0.0%
Sema.NumUnloadedLazyIterableDeclContexts | 6,349,199 | 6,327,323 | -21,876 | -0.34%
Sema.OpaqueReadOwnershipRequest | 282,386 | 282,386 | 0 | 0.0%
Sema.OverriddenDeclsRequest | 2,423,807 | 2,446,809 | 23,002 | 0.95%
Sema.PropertyWrapperBackingPropertyInfoRequest | 510,547 | 510,547 | 0 | 0.0%
Sema.PropertyWrapperBackingPropertyTypeRequest | 515,933 | 515,933 | 0 | 0.0%
Sema.PropertyWrapperMutabilityRequest | 621,980 | 621,980 | 0 | 0.0%
Sema.PropertyWrapperTypeInfoRequest | 1 | 1 | 0 | 0.0%
Sema.ProtocolRequiresClassRequest | 92,448 | 92,322 | -126 | -0.14%
Sema.RangeInfoRequest | 0 | 0 | 0 | 0.0%
Sema.RequirementRequest | 112,203 | 112,218 | 15 | 0.01%
Sema.RequirementSignatureRequest | 762,221 | 762,869 | 648 | 0.09%
Sema.RequiresOpaqueAccessorsRequest | 1,325,476 | 1,325,476 | 0 | 0.0%
Sema.RequiresOpaqueModifyCoroutineRequest | 273,633 | 273,633 | 0 | 0.0%
Sema.ResilienceExpansionRequest | 1,843,120 | 1,843,186 | 66 | 0.0%
Sema.ResolveProtocolNameRequest | 0 | 0 | 0 | 0.0%
Sema.RootAndResultTypeOfKeypathDynamicMemberRequest | 0 | 0 | 0 | 0.0%
Sema.RootTypeOfKeypathDynamicMemberRequest | 0 | 0 | 0 | 0.0%
Sema.SelfAccessKindRequest | 6,384,521 | 6,374,231 | -10,290 | -0.16%
Sema.SelfBoundsFromWhereClauseRequest | 7,990,515 | 8,009,494 | 18,979 | 0.24%
Sema.SetterAccessLevelRequest | 151,118 | 151,118 | 0 | 0.0%
Sema.StorageImplInfoRequest | 1,209,542 | 1,209,542 | 0 | 0.0%
Sema.StoredPropertiesAndMissingMembersRequest | 32,443 | 32,443 | 0 | 0.0%
Sema.StoredPropertiesRequest | 350,233 | 350,233 | 0 | 0.0%
Sema.StructuralTypeRequest | 0 | 0 | 0 | 0.0%
Sema.SuperclassDeclRequest | 498,581 | 499,354 | 773 | 0.16%
Sema.SuperclassTypeRequest | 56,563 | 56,563 | 0 | 0.0%
Sema.SynthesizeAccessorRequest | 296,229 | 296,229 | 0 | 0.0%
Sema.TypeCheckFunctionBodyUntilRequest | 527,256 | 527,256 | 0 | 0.0%
Sema.TypeDeclsFromWhereClauseRequest | 29,943 | 29,954 | 11 | 0.04%
Sema.TypeRelationCheckRequest | 0 | 0 | 0 | 0.0%
Sema.UnderlyingTypeDeclsReferencedRequest | 293,298 | 294,852 | 1,554 | 0.53%
</details>

<a name="release-49cfcd52a955fa2d">

# Release

<a name="release-brief-49cfcd52a955fa2d">

## release brief
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (3)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
Frontend.NumInstructionsExecuted | 47,331,383,660,595 | 47,300,754,530,027 | -30,629,130,568 | -0.06%
LLVM.NumLLVMBytesOutput | 1,530,048,510 | 1,530,055,474 | 6,964 | 0.0%
time.swift-driver.wall | 8623.2s | 8607.7s | -15.5s | -0.18%
</details>

<a name="release-detailed-49cfcd52a955fa2d">

## release detailed
<details>
<summary>Regressed (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Improved (0)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
</details>
<details>
<summary>Unchanged (delta < 1.0% or delta < 100.0ms) (21)</summary>

name | old | new | delta | delta_pct
---: | ---: | ---: | ---: | ---:
AST.NumLoadedModules | 30,333 | 30,333 | 0 | 0.0%
AST.NumTotalClangImportedEntities | 1,231,555 | 1,231,555 | 0 | 0.0%
IRModule.NumIRBasicBlocks | 5,608,688 | 5,608,688 | 0 | 0.0%
IRModule.NumIRFunctions | 2,849,076 | 2,849,076 | 0 | 0.0%
IRModule.NumIRGlobals | 3,030,523 | 3,030,523 | 0 | 0.0%
IRModule.NumIRInsts | 52,605,316 | 52,605,316 | 0 | 0.0%
IRModule.NumIRValueSymbols | 5,523,291 | 5,523,291 | 0 | 0.0%
LLVM.NumLLVMBytesOutput | 1,530,048,510 | 1,530,055,474 | 6,964 | 0.0%
SILModule.NumSILGenFunctions | 1,191,481 | 1,191,481 | 0 | 0.0%
SILModule.NumSILOptFunctions | 1,711,269 | 1,711,269 | 0 | 0.0%
Sema.NumConformancesDeserialized | 4,079,694 | 4,079,694 | 0 | 0.0%
Sema.NumConstraintScopes | 26,612,735 | 26,612,735 | 0 | 0.0%
Sema.NumDeclsDeserialized | 10,411,974 | 10,411,974 | 0 | 0.0%
Sema.NumDeclsValidated | 1,826,150 | 1,826,150 | 0 | 0.0%
Sema.NumFunctionsTypechecked | 530,878 | 530,878 | 0 | 0.0%
Sema.NumGenericSignatureBuilders | 279,409 | 279,409 | 0 | 0.0%
Sema.NumLazyGenericEnvironments | 1,972,770 | 1,972,770 | 0 | 0.0%
Sema.NumLazyGenericEnvironmentsLoaded | 34,177 | 34,177 | 0 | 0.0%
Sema.NumLazyIterableDeclContexts | 1,326,390 | 1,326,390 | 0 | 0.0%
Sema.NumTypesDeserialized | 5,675,438 | 5,675,438 | 0 | 0.0%
Sema.NumTypesValidated | 947,067 | 947,067 | 0 | 0.0%
</details>

slavapestov commented:
The increase in AST context bytes allocated is expected. We allocate some things in the permanent arena that used to be per-SILModule.

jrose-apple commented:
A little disappointed that neither the peak memory use nor the number of instructions executed seemed to improve significantly in the debug-batch configuration, but maybe the projects don't have enough files per target for it to show up.

slavapestov commented:
apple/swift-lldb#1930
@swift-ci Please smoke test

jrose-apple commented:
Oh, hm. I wonder if it was done the previous way to support the deallocation of the ASTContext before processing the LLVM IR. As long as we didn't disrupt that we should be okay.
